File: examples/api/python/denpie_client.py
#!/usr/bin/env python3
"""Small dependency-light Denpie API v1 client."""

from __future__ import annotations

import os
import sys
import uuid
from urllib.error import HTTPError
from urllib.request import Request, urlopen

import denpie_pb2 as pb


def api_call(envelope: pb.ApiV1Request) -> pb.ApiV1Response:
    endpoint = os.environ.get("DENPIE_URL", "http://127.0.0.1:3017/api/v1")
    headers = {"Content-Type": "application/x-protobuf"}
    api_key = os.environ.get("DENPIE_API_KEY")
    if api_key:
        headers["Authorization"] = f"Bearer {api_key}"
    request = Request(
        endpoint,
        data=envelope.SerializeToString(),
        headers=headers,
        method="POST",
    )
    try:
        with urlopen(request, timeout=30) as response:
            status = response.status
            body = response.read()
    except HTTPError as error:
        status = error.code
        body = error.read()

    result = pb.ApiV1Response()
    result.ParseFromString(body)
    if result.WhichOneof("outcome") == "error":
        code = pb.ApiErrorCode.Name(result.error.code)
        raise RuntimeError(
            f"Denpie returned HTTP {status} {code}: {result.error.message} "
            f"(request_id={result.request_id}, retryable={result.error.retryable})"
        )
    if status < 200 or status >= 300 or result.WhichOneof("outcome") != "success":
        raise RuntimeError(f"invalid Denpie response: HTTP {status}")
    return result


def get_info() -> pb.ApiV1Response:
    return api_call(
        pb.ApiV1Request(
            request_id="python-get-api-info",
            call=pb.ApiRequest(get_api_info=pb.Empty()),
        )
    )


def list_cards() -> pb.ApiV1Response:
    return api_call(
        pb.ApiV1Request(
            request_id="python-list-flow-cards",
            call=pb.ApiRequest(
                list_flow_cards=pb.ListFlowCardsRequest(page_size=12)
            ),
        )
    )


def create_document(idempotency_key: str) -> pb.ApiV1Response:
    return api_call(
        pb.ApiV1Request(
            request_id="python-create-document",
            idempotency_key=idempotency_key,
            call=pb.ApiRequest(
                create_document=pb.AddDocumentRequest(
                    source_type="document",
                    title="Python API example",
                    content="Created by the checked-in Python client example.",
                )
            ),
        )
    )


def self_test() -> None:
    request = pb.ApiV1Request(
        request_id="python-self-test",
        call=pb.ApiRequest(get_api_info=pb.Empty()),
    )
    decoded = pb.ApiV1Request.FromString(request.SerializeToString())
    assert decoded.call.WhichOneof("op") == "get_api_info"
    print("Python protobuf client self-test passed")


def main() -> int:
    command = sys.argv[1] if len(sys.argv) > 1 else "info"
    if command == "--self-test":
        self_test()
        return 0
    if command == "info":
        response = get_info()
    elif command == "cards":
        response = list_cards()
    elif command == "create-document":
        key = os.environ.get("DENPIE_IDEMPOTENCY_KEY", str(uuid.uuid4()))
        print(f"idempotency_key={key}", file=sys.stderr)
        response = create_document(key)
    else:
        print("usage: denpie_client.py [info|cards|create-document|--self-test]", file=sys.stderr)
        return 2
    print(response)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: examples/api/typescript/denpie-client.ts
import { randomUUID } from "node:crypto";
import path from "node:path";
import { fileURLToPath } from "node:url";
import protobuf from "protobufjs";

const here = path.dirname(fileURLToPath(import.meta.url));
const schemaPath = process.env.DENPIE_SCHEMA ?? path.resolve(here, "../../../proto/denpie.proto");
const root = await protobuf.load(schemaPath);
const ApiV1Request = root.lookupType("denpie.ApiV1Request");
const ApiV1Response = root.lookupType("denpie.ApiV1Response");

type Payload = Record<string, unknown>;

async function apiCall(payload: Payload): Promise<Record<string, unknown>> {
  const validationError = ApiV1Request.verify(payload);
  if (validationError) throw new Error(`invalid request: ${validationError}`);

  const encoded = ApiV1Request.encode(ApiV1Request.create(payload)).finish();
  const headers: Record<string, string> = {
    "Content-Type": "application/x-protobuf",
  };
  if (process.env.DENPIE_API_KEY) {
    headers.Authorization = `Bearer ${process.env.DENPIE_API_KEY}`;
  }
  const response = await fetch(
    process.env.DENPIE_URL ?? "http://127.0.0.1:3017/api/v1",
    { method: "POST", headers, body: Buffer.from(encoded) },
  );
  const bytes = new Uint8Array(await response.arrayBuffer());
  const decoded = ApiV1Response.toObject(ApiV1Response.decode(bytes), {
    enums: String,
    longs: String,
    oneofs: true,
  }) as Record<string, unknown>;
  if (!response.ok || decoded.outcome === "error") {
    throw new Error(`Denpie returned HTTP ${response.status}: ${JSON.stringify(decoded)}`);
  }
  return decoded;
}

function infoRequest(): Payload {
  return {
    requestId: "typescript-get-api-info",
    call: { getApiInfo: {} },
  };
}

function cardsRequest(): Payload {
  return {
    requestId: "typescript-list-flow-cards",
    call: { listFlowCards: { pageSize: 12 } },
  };
}

function createDocumentRequest(idempotencyKey: string): Payload {
  return {
    requestId: "typescript-create-document",
    idempotencyKey,
    call: {
      createDocument: {
        sourceType: "document",
        title: "TypeScript API example",
        content: "Created by the checked-in TypeScript client example.",
      },
    },
  };
}

function selfTest(): void {
  const encoded = ApiV1Request.encode(ApiV1Request.create(infoRequest())).finish();
  const decoded = ApiV1Request.toObject(ApiV1Request.decode(encoded), { oneofs: true });
  if (decoded.call?.op !== "getApiInfo") throw new Error("oneof did not round-trip");
  console.log("TypeScript protobuf client self-test passed");
}

const command = process.argv[2] ?? "info";
if (command === "--self-test") {
  selfTest();
} else if (command === "info") {
  console.log(await apiCall(infoRequest()));
} else if (command === "cards") {
  console.log(await apiCall(cardsRequest()));
} else if (command === "create-document") {
  const idempotencyKey = process.env.DENPIE_IDEMPOTENCY_KEY ?? randomUUID();
  console.error(`idempotency_key=${idempotencyKey}`);
  console.log(await apiCall(createDocumentRequest(idempotencyKey)));
} else {
  console.error("usage: denpie-client.ts [info|cards|create-document|--self-test]");
  process.exitCode = 2;
}
